### ISSUE TITLE
Unicode problem.

### ISSUE THREAD

**Issue by zaksnet** [godotengine/godot#30396]
**Godot version:**
3.1 Stable (tested with latest master also)

**OS/device including version:**
Windows 10

**Issue description:**
I am receiving unicode messages through sockets that contain unicode characters, in my case Greek language characters.
- StreamPeer `get_utf8_string` gives me `Unicode error: invalid skip`.
- PoolByteArray `get_string_from_utf8` gives me `Unicode error: invalid skip`.
![Screenshot_3](https://user-images.githubusercontent.com/1850856/60765814-d0d3c480-a0a8-11e9-9a3a-307323554fde.png)

- PoolByteArray `get_string_from_ascii` give me the full message but all unicode characters are converted to questionmarks.
![Screenshot_4](https://user-images.githubusercontent.com/1850856/60765818-f8c32800-a0a8-11e9-85e3-37bf219870fe.png)

I'm a bit frustrated with thisone, searched through issues but could not find any solution.

**Comment by raymoo** [comment]
- How are strings formatted on the socket? Do they have a length prepended or are they nul-terminated? Are you sure that it's UTF-8 and not another encoding like UTF-16?
- What argument (if any) are you passing to `get_utf8_string`? If you pass no length argument, the function expects to first read the length of the string and then read that many bytes.
- What does `get_string_from_utf8` return when it gives that error?

**Comment by zaksnet** [comment]
> How are strings formatted on the socket? Do they have a length prepended or are they nul-terminated? Are you sure that it's UTF-8 and not another encoding like UTF-16?

nul-terminated in my case. I'm not sure if its UTF-16 since i have not written the server part. Unfortunately, i don't know what kind of encoding the server uses but, since this is the most probable case, any idea how would i proceed with decoding UTF-16 in Godot?

> What argument (if any) are you passing to get_utf8_string? If you pass no length argument, the function expects to first read the length of the string and then read that many bytes.

```
var bytes = client.get_available_bytes()
var message = client.get_utf8_string('bytes')
```

> What does get_string_from_utf8 return when it gives that error?

Returns the part of the message that was encoded until the error occurred. If the message was 500 chars and the error occurred at char 300 then the result is 299 chars. Everything beyond that is skipped.

**Comment by raymoo** [comment]
I actually think UTF-16 is unlikely, I was asking just in case. It may also be a non-unicode greek encoding, but I'm assuming you know it's some kind of unicode since that's what the issue title says. If you're not sure, then I think a non-unicode encoding is more likely than UTF-16, but I'm not really that familiar.

Also you shouldn't rely on a string after `get_available_bytes` to be exactly one message. That function just tells you how much data it has received so far, and it's possible that it has received more or less than one full message. If you read that number of bytes you might not have a full message, or you might have a full message and the beginning of another message. Unicode characters can be multiple bytes, so the partial data might even be split in the middle of a character.

To handle null-terminated strings you would need to:
- Keep a `PoolByteArray` to hold incomplete strings
- When you read data from the client, use `get_data` instead of `get_utf8_string`, because the data can end in the middle of a unicode character, and then you would have an invalid character and lose that data.
- If there's a null character in the data, then add everything before the null character to the `PoolByteArray` and then turn that whole thing into a string, and then clear the buffer. Put everything after the null character into the cleared `PoolByteArray` since that is the start of the next string.
- If there is no null character, just add the whole thing to the buffer because you haven't completed a string yet.

**Comment by zaksnet** [comment]
> I actually think UTF-16 is unlikely, I was asking just in case. It may also be a non-unicode greek encoding, but I'm assuming you know it's some kind of unicode since that's what the issue title says. If you're not sure, then I think a non-unicode encoding is more likely than UTF-16, but I'm not really that familiar.
> 
> Also you shouldn't rely on a string after `get_available_bytes` to be exactly one message. That function just tells you how much data it has received so far, and it's possible that it has received more or less than one full message. If you read that number of bytes you might not have a full message, or you might have a full message and the beginning of another message. Unicode characters can be multiple bytes, so the partial data might even be split in the middle of a character.
> 
> To handle null-terminated strings you would need to:
> 
> * Keep a `PoolByteArray` to hold incomplete strings
> * When you read data from the client, use `get_data` instead of `get_utf8_string`, because the data can end in the middle of a unicode character, and then you would have an invalid character and lose that data.
> * If there's a null character in the data, then add everything before the null character to the `PoolByteArray` and then turn that whole thing into a string, and then clear the buffer. Put everything after the null character into the cleared `PoolByteArray` since that is the start of the next string.
> * If there is no null character, just add the whole thing to the buffer because you haven't completed a string yet.

Sorry, my fault, all messages end with an end tag ({end} in my case) not null terminated. Of course, everything you said applies in this case but, i'm just mentioning it because i have already build a mechanism to detect when the full message is received.

**Comment by Faless** [comment]
@zaksnet how do you send the string? Can you provide a minimal reproduction project? How do you detect the end tag?

**Comment by zaksnet** [comment]
> @zaksnet how do you send the string? Can you provide a minimal reproduction project? How do you detect the end tag?

The string is being sent from another application (of which i do not currently have the source code). I am doing:
```
const ENDTAG = "{end}"
	if is_connected_to_host:
		var bytes = client.get_available_bytes()
		if bytes > 0:
			var message: String = client.get_string(bytes)
			var messages = message.split(ENDTAG)
```
If a split does not end with ENDTAG, i add it to the next message.
1) I dont think this is network related. It just happend that the data i get came from the network (i think the same thing would happend if i loaded them from a file).
2) I am guessing `get_string` decodes the message to ascii since i get the same result as `get_string_from_ascii`. That is, the whole string is fine (i'm pretty sure there is not partial data) except the unicde parts which are represented as questionmarks.

In conclusion, i think this is a Unicode decoding problem and not network related.

**Comment by Anutrix** [comment]
Can you test it on windows to make sure it's not platform specific? Because it seems the issue is same as #30409.

**Comment by Faless** [comment]
> If a split does not end with ENDTAG, i add it to the next message.

Okay, there lies the problem, unicode is variably one or two bytes.

You **can't** decode partial data.
When reading from the stream you **MUST** detect the end tag, and only decode **after** you have the full message, as @raymoo explained you above.

**Comment by zaksnet** [comment]
> > If a split does not end with ENDTAG, i add it to the next message.
> 
> Okay, there lies the problem, unicode is variably one or two bytes.
> 
> You **can't** decode partial data.
> When reading from the stream you **MUST** detect the end tag, and only decode **after** you have the full message, as @raymoo explained you above.

Cool, thing is, i never had to add data to the next message so far. Messages are small and this is being tested locally so i get the whole string all the time, some times i get 3,4 messages at the same time and all of these might contain Unicode data so between endtag 1 and endtag 2 there are question marks (this is just one stream of data) and there are 4 endtags so i don't see how the data could be malformed in my case.

**Comment by Faless** [comment]
@zaksnet can you give us the raw bytes of one message so we can try it?

**Comment by zaksnet** [comment]
Sure, check the attached file. I used File > store_buffer to store the PoolByteArray in the file and all text seems to be fine in there.
[unicode.txt](https://github.com/godotengine/godot/files/3367487/unicode.txt)

**Comment by Faless** [comment]
> EDIT: nvm, you want the raw bytes, how can i send you this?

`print(Array(client.get_data(bytes)[1]))`

**Comment by zaksnet** [comment]
That only prints `[0, [PoolByteArray]]`

**Comment by zaksnet** [comment]
```
[85, 115, 101, 114, 83, 101, 116, 73, 110, 102, 111, 114, 109, 97, 116, 105, 111, 110, 61, 122, 97, 107, 94, 122, 97, 107, 94, 49, 49, 47, 49, 49, 47, 49, 57, 56, 48, 32, 49, 50, 58, 48, 48, 58, 48, 48, 32, 240, 236, 94, 122, 97, 107, 94, 94, 49, 50, 51, 94, 94, 94, 122, 97, 107, 115, 99, 111, 109, 112, 117, 116, 101, 114, 115, 64, 104, 111, 116, 109, 97, 105, 108, 46, 99, 111, 109, 94, 49, 50, 48, 53, 54, 49, 49, 56, 49, 54, 94, 94, 49, 94, 94, 48, 94, 56, 47, 55, 47, 50, 48, 49, 57, 32, 49, 50, 58, 51, 50, 58, 49, 57, 32, 236, 236, 94, 204, 221, 235, 231, 94, 94, 50, 94, 50, 54, 50, 51, 94, 50, 51, 52, 55, 52, 94, 94, 123, 101, 110, 100, 125, 85, 115, 101, 114, 76, 111, 103, 105, 110, 65, 112, 112, 114, 111, 118, 101, 100, 61, 122, 97, 107, 94, 83, 104, 118, 86, 75, 106, 113, 107, 102, 113, 122, 90, 104, 83, 53, 108, 110, 70, 82, 47, 74, 99, 68, 52, 83, 50, 52, 61, 94, 204, 221, 235, 231, 123, 101, 110, 100, 125, 83, 101, 116, 83, 101, 99, 117, 114, 105, 116, 121, 65, 108, 108, 111, 119, 101, 100, 68, 114, 105, 118, 101, 115, 61, 49, 48, 52, 56, 55, 48, 53, 123, 101, 110, 100, 125, 83, 101, 116, 83, 101, 99, 117, 114, 105, 116, 121, 83, 101, 116, 116, 105, 110, 103, 115, 61, 49, 53, 59, 50, 53, 59, 51, 51, 59, 51, 52, 59, 54, 50, 59, 49, 59, 50, 59, 51, 59, 52, 59, 54, 59, 55, 59, 56, 59, 49, 51, 59, 49, 52, 59, 50, 52, 59, 50, 56, 59, 51, 49, 59, 51, 50, 59, 49, 54, 59, 49, 56, 59, 49, 57, 59, 50, 48, 59, 50, 49, 59, 50, 50, 59, 50, 51, 59, 50, 54, 59, 50, 55, 59, 51, 54, 59, 51, 55, 59, 51, 56, 59, 51, 57, 59, 52, 48, 59, 52, 49, 59, 52, 50, 59, 52, 51, 59, 52, 52, 59, 52, 53, 59, 52, 55, 59, 52, 56, 59, 52, 57, 59, 53, 48, 59, 53, 50, 59, 54, 49, 59, 54, 51, 59, 54, 52, 59, 54, 53, 59, 53, 52, 59, 123, 101, 110, 100, 125, 83, 101, 116, 83, 101, 99, 117, 114, 105, 116, 121, 65, 112, 112, 108, 105, 99, 97, 116, 105, 111, 110, 70, 105, 108, 116, 10[...]
[output overflow, print less text!]
```
[bytes.txt](https://github.com/godotengine/godot/files/3367711/bytes.txt)

**Comment by Faless** [comment]
The data seems to be invalid UTF-8:
```
iconv -f UTF-8 unicode.txt -o /dev/null; echo $?
iconv: illegal input sequence at position 47
1
```

**Comment by Faless** [comment]
Specifically, the `ð` is encoded as `240` decimal, which [seems to be UTF-16](http://www.fileformat.info/info/unicode/char/f0/index.htm)

**Comment by Faless** [comment]
`PoolByteArray([85, 115, 101, 114, 83, 101, 116, 73, 110, 102, 111, 114, 109, 97, 116, 105, 111, 110, 61, 122, 97, 107, 94, 122, 97, 107, 94, 49, 49, 47, 49, 49, 47, 49, 57, 56, 48, 32, 49, 50, 58, 48, 48, 58, 48, 48, 32,`**240**`, 236, 94, 122])`

**Comment by Faless** [comment]
If the letter there is supposed to be `π` and not `ð` then the text is not UTF-8 encoded but `ISO/IEC 8859-7`

**Comment by zaksnet** [comment]
> If the letter there is supposed to be `π` and not `ð` then the text is not UTF-8 encoded but `ISO/IEC 8859-7` (or `Windows-1253`)

Cool! Yes, it is a `π`. A lot to consume here but, how would i proceed in that case? Can i somehow decode this properly?

**Comment by zaksnet** [comment]
Thank you for researching this BTW! Appreciated!

**Comment by Faless** [comment]
> Can i somehow decode this properly?

I don't think Godot supports other encodings beside UTF-8 and ASCII.
And really, no one should use non UTF-8 encodings at this point in time.

Check if the service from which you are getting the data supports utf-8 encoding (maybe with an extra query parameter).

The only other way I can thing of, is using a translation table from `ISO/IEC 8859-7` to `UTF-8` and applying it to the raw bytes, then using `get_string_from_utf8` on the converted bytes.

**Comment by LinuxUserGD** [comment]
similar, but offtopic to this issue:
@Faless
Did you see this PR?
https://github.com/Faless/godot-websocket-demo/pull/3
It is for Godot 3.1+ support and changes the encoding in your example project.
Could you please either merge or reject it?
